src/Memory/include/Memory/MemoryInfoProducer.h

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_MEMORY_INFO_PRODUCER_H_
+#define ORBIT_MEMORY_INFO_PRODUCER_H_
+
+#include <atomic>
+#include <thread>
+
+#include "Memory/MemoryInfoListener.h"
+#include "capture.pb.h"
+
+namespace orbit_memory {
+class MemoryInfoProducer {
+ public:
+  explicit MemoryInfoProducer(uint64_t memory_sampling_period_ns)
+      : sampling_period_ns_(memory_sampling_period_ns) {}
+
+  MemoryInfoProducer(const MemoryInfoProducer&) = delete;
+  MemoryInfoProducer& operator=(const MemoryInfoProducer&) = delete;
+  MemoryInfoProducer(MemoryInfoProducer&&) = delete;
+  MemoryInfoProducer& operator=(MemoryInfoProducer&&) = delete;
+
+  ~MemoryInfoProducer() { Stop(); }
+
+  void SetListener(MemoryInfoListener* listener) { listener_ = listener; }
+
+  void Start();
+  void Stop();
+
+ private:
+  void Run();
+  void ProduceTotalMemoryUsageInfoAndSendToListener();
+
+  uint64_t sampling_period_ns_;
+  MemoryInfoListener* listener_ = nullptr;
+  std::shared_ptr<std::atomic<bool>> exit_requested_ = std::make_unique<std::atomic<bool>>(true);
+  std::shared_ptr<std::thread> thread_;
```

Comment:
Do you need this to be wrapped inside a smart pointer?

Reply:
Changed to use `std::thread* thread_;` Thanks for pointing out this.

Reply:
But now we are creating the thread with `new`, which seems very suspicious to me. Couldn't it be just a member of the producer? E.g. as `std::optional`, if you need to "reset" it on capture stop?
If not it should be probably still a `unique_ptr`.

But passing `std::thread*` is better than the shared_ptr, I agree.